api/v3/GreenpeaceMessageTemplate/Send.php

```diff
@@ -0,0 +1,65 @@
+<?php
+use CRM_Uimods_ExtensionUtil as E;
+
+/**
+ * GreenpeaceMessageTemplate.Send API specification (optional)
+ * This is used for documentation and validation.
+ *
+ * @param array $spec description of fields supported by this API call
+ *
+ * @see https://docs.civicrm.org/dev/en/latest/framework/api-architecture/
+ */
+function _civicrm_api3_greenpeace_message_template_send_spec(&$spec) {
+  $spec['id'] = [
+    'name'         => 'id',
+    'title'        => 'MessageTemplate ID',
+    'type'         => CRM_Utils_Type::T_INT,
+    'api.required' => 1
+  ];
+
+  $spec['to_emails'] = [
+    'name'         => 'to_emails',
+    'title'        => 'To Emails (comma separated list)',
+    'type'         => CRM_Utils_Type::T_STRING,
+    'api.required' => 1
+  ];
+}
+
+/**
+ * GreenpeaceMessageTemplate.Send API
+ *
+ * @param array $params
+ *
+ * @return array
+ *   API result descriptor
+ *
+ * @see civicrm_api3_create_success
+ *
+ * @throws API_Exception
+ */
+function civicrm_api3_greenpeace_message_template_send($params) {
+
+  $return_values = [
+    'valid' => [],
+    'invalid' => []
+  ];
+
+  try {
+    $emails = explode(",", $params['to_emails']);
+    foreach ($emails as $email) {
+      if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+        $return_values['invalid'][] = $email;
+      } else {
+        $message_template = civicrm_api3('MessageTemplate', 'send', [
```

Comment:
We can pass through `$params` as a whole (except for `to_email`, which we set to `$email`) so we support all other parameters supported by `MessageTemplate.send` as well. (They don't have to be defined in `_civicrm_api3_greenpeace_message_template_send_spec`, they'll just magically work if someone provides values). For example, we sometimes also use `template_params`.

Reply:
commit e38c4661d58158066094fc5a5c371e41a48b8706